### PR TITLE
Support non-existing file mock and coerce it into a file on write bas…

### PR DESCRIPTION
**Pull request by toddr**
…ed open.

We're going to clear mode now on unlink since mode is meaningless when there's
no existing thing at the mocked location.

This also exposed several bad assumptions in tests.

### CODE REVIEW

File: lib/Test/MockFile.pm [hunk 1]
```diff
@@ -821,13 +821,9 @@ sub dir {
...
 
-    my $perms = S_IFPERMS & 0777;
-    my %stats = ( 'mode' => ( $perms ^ umask ) | S_IFDIR );
-
-    # TODO: Add stat information
+    my %stats;
+    my $has_content = 0;
 
-    # FIXME: Quick and dirty: provide a helper method?
-    my $has_content = grep m{^\Q$path/\E}xms, %files_being_mocked;
```
Comment (toddr) [inline lib/Test/MockFile.pm:830]: I broke shit.

Comment (toddr) [reply lib/Test/MockFile.pm:830]: @xsawyerx disagrees with this.

File: lib/Test/MockFile.pm [hunk 2]
```diff
@@ -1206,19 +1202,13 @@ sub contents {
...
-    if ( $self->is_file() ) {
-        if ( defined $new_contents ) {
-            ref $new_contents
-              and confess('File contents must be a simple string');
-
-            # XXX Why use $_[1] directly?
-            $self->{'contents'} = $_[1];
-        }
-
+    if ( !defined $new_contents ) {
```
Comment (toddr) [inline lib/Test/MockFile.pm:1205]: $self->is_file && !defined $new_contents

File: lib/Test/MockFile.pm [hunk 3]
```diff
@@ -1379,6 +1372,8 @@ returns true/false, depending on whether this object is a symlink.
 sub is_link {
     my ($self) = @_;
 
+    $self->{'mode'} // return 0;
```
Comment (toddr) [inline lib/Test/MockFile.pm:1375]: `$self->{'mode'} || return 0;`

File: lib/Test/MockFile.pm [hunk 4]
```diff
@@ -1391,6 +1386,8 @@ returns true/false, depending on whether this object is a directory.
 sub is_dir {
     my ($self) = @_;
 
+    $self->{'mode'} // return 0;
```
Comment (toddr) [inline lib/Test/MockFile.pm:1389]: `$self->{'mode'} || return 0;`

File: lib/Test/MockFile.pm [hunk 5]
```diff
@@ -1403,6 +1400,8 @@ returns true/false, depending on whether this object is a regular file.
 sub is_file {
     my ($self) = @_;
 
+    $self->{'mode'} // return 0;
```
Comment (toddr) [inline lib/Test/MockFile.pm:1403]: `$self->{'mode'} || return 0;`

File: lib/Test/MockFile.pm [hunk 6]
```diff
@@ -1770,9 +1769,31 @@ sub __open (*;$@) {
...
+                return;
+            }
+            elsif ( $mock_file->is_link ) {
+                die("Failed to follow link. We shouldn't have gotten here.");
+            }
+            else {
+                die("Unhandled mocked open for a non-file");
+            }
+        }
+        elsif ( defined $mode && substr( $mode, 0, 1 ) eq '>' ) {
```
Comment (toddr) [inline lib/Test/MockFile.pm:1792]: What about `+>` ?

File: lib/Test/MockFile.pm (hunk 5 above)
Comment (toddr) [reply lib/Test/MockFile.pm:1403]: `$self->exists;`

File: lib/Test/MockFile.pm [hunk 7]
```diff
@@ -642,7 +642,7 @@ sub file {
     }
 
     my $perms = S_IFPERMS & ( defined $stats{'mode'} ? int( $stats{'mode'} ) : 0666 );
-    $stats{'mode'} = ( $perms ^ umask ) | S_IFREG;
+    $stats{'mode'} = ( $perms & ~umask ) | S_IFREG;
```
Comment (toddr) [inline lib/Test/MockFile.pm:645]: `_make_mode('0755', S_IFREG);`

File: lib/Test/MockFile.pm (hunk 6 above)
Comment (toddr) [reply lib/Test/MockFile.pm:1802]: @xsawyerx what about:
```perl
elsif ( defined $mode && $mode =~ tr/>+// ) { # If this is an open for write.
```

File: lib/Test/MockFile.pm (hunk 7 above)
Comment (toddr) [reply lib/Test/MockFile.pm:656]: done.

File: lib/Test/MockFile.pm (hunk 6 above)
Comment (xsawyerx) [reply lib/Test/MockFile.pm]: No objections. I keep forgetting about `tr///`. Would you like me to update it?

File: lib/Test/MockFile.pm (hunk 1 above)
Comment (xsawyerx) [reply lib/Test/MockFile.pm:841]: I imagine this was resolved?

**Pull request closed without merging**